### ISSUE TITLE
High background CPU usage after switching projects many times

### ISSUE THREAD

**Issue by peppy** [jesseduffield/lazygit#616]
The more times you switch projects (1 - enter - enter), the higher the idle CPU usage gets. Maybe a file system watcher or background poll not being cleaned up correctly.

![CleanShot 2020-01-31 at 16 42 36](https://user-images.githubusercontent.com/191335/73521190-b8face00-4448-11ea-9887-e83e1c6ad52f.png)

Restarting lazygit will fix this temporarily.

[Sample of lazygit.txt](https://github.com/jesseduffield/lazygit/files/4138113/Sample.of.lazygit.txt)

**Comment by jesseduffield** [comment]
I've merged a PR  https://github.com/jesseduffield/lazygit/pull/617that will make some headway towards solving the problem, though testing locally I'm still seeing an increase in CPU as I switch repos. As I stated in that PR, looks like it's mostly due to threads waiting on mutexes, and I'm not too sure how to find out where those waits are happening. Seems it will require some more investigation.

**Comment by jesseduffield** [comment]
update: go1.13 doesn't do much better than go1.12. Seems like the CPU I'm getting now gets somewhat high regardless of whether I'm switching repos or not

**Comment by jesseduffield** [comment]
using StackImpact I've narrowed it down to a couple of things: 
![image](https://user-images.githubusercontent.com/8456633/73532121-20784400-4470-11ea-822a-e4933a10301a.png)

now I just need to find a way to still do things periodically but not have it waste thread time. I know with the loaderTick function that only really needs to be executing when a loader is displaying so I can find a way around that. Not so sure with the other things

**Comment by jesseduffield** [comment]
on further inspection this really just comes down to the two timer loops that have 50ms ticks, both of which are for the sake of rendering a spinning loading character (and therefore don't need to be running all the time). So if I can find a way to only run those loops when they're needed, that will solve the problem

**Comment by peppy** [comment]
Thanks for the changes! Unfortunately, on the latest master I'm seeing even worse CPU usage. It looks to be steadily rising the longer I have lazygit running.

![CleanShot 2020-01-31 at 20 36 43](https://user-images.githubusercontent.com/191335/73536525-6e3d7e00-4469-11ea-8bd2-5585c396e561.png)

Let me know if you need further sampling or otherwise from my end.

**Comment by jesseduffield** [comment]
I've just merged another PR, I think it should be fixed for good this time: https://github.com/jesseduffield/lazygit/pull/618

**Comment by peppy** [comment]
Looking much better now, thanks!

**Comment by peppy** [comment]
I managed to repro highish CPU after some hours of use. Do you have a preferred way for me to sample the process (apart from macOS sampling)?

**Comment by jesseduffield** [comment]
@peppy I put up another PR semi-recently that addressed further CPU issues, would you be able to test out the latest release and see if the issue persists?

**Comment by jesseduffield** [comment]
otherwise the sampling approach you used in the OP is fine

**Comment by peppy** [comment]
I've already updated to latest master earlier today, will let you know if it occurs again!

**Comment by jesseduffield** [comment]
cool :)

**Comment by peppy** [comment]
Managed to get it stuck at using one CPU:
[Sample of lazygit.txt](https://github.com/jesseduffield/lazygit/files/4146047/Sample.of.lazygit.txt)

**Comment by jesseduffield** [comment]
Ah I think I know what the issue is. I'll have a closer look after work

**Comment by jesseduffield** [comment]
found a bug that fixed the issue on my end, PR is https://github.com/jesseduffield/lazygit/pull/624

Would you be able to test again on the latest master?

**Comment by peppy** [comment]
So far so good! Will continue to test for another day of usage and then close this issue again.

**Comment by peppy** [comment]
Looks fixed for good!

**Comment by jesseduffield** [comment]
woohoo!
